### PR TITLE
[SV] Remove if/ifdef/always/initial op if they are empty

### LINKED ISSUE #636: canonicalize 'sv.if true' and 'sv.if false' away
Particularly after recursive simplification it is common to get dead conditions.  I see this in one random IR dump:

```
   sv.always posedge %clock  {
      sv.ifdef.procedural "SYNTHESIS"  {
      } else  {
        sv.if %false  {
          sv.fatal
        }
      }
    }
    sv.always posedge %clock  {
      sv.ifdef.procedural "SYNTHESIS"  {
      } else  {
        sv.if %false  {
          sv.fwrite "Assertion failed: 'A' channel carries an address illegal for the specified bank visibility\0A    at Monitor.scala:81 assert (visible(edge.address(bundle), bundle.source, edge), \22'A' channel carries an address illegal for the specified bank visibility\22)\0A"
        }
      }
    }
```

(the sv.always nodes aren't getting merged because lower2rtl is generating them instead of alwaysff).

### PR DESCRIPTION
**Pull request by uenoku**
WIll close #636.

### CODE REVIEW

**Review by darthscsi — APPROVED** [review]

File: lib/Dialect/SV/SVOps.cpp [hunk 1]
```diff
@@ -166,6 +166,33 @@ void IfDefOp::build(OpBuilder &odsBuilder, OperationState &result,
   }
 }
 
+static bool isEmptyBlockExceptForTerminator(Block *block) {
+  if (!block)
+    return true;
```
Comment (lattner) [inline lib/Dialect/SV/SVOps.cpp:171]: I don't think you need this check for any caller, make it an assert?

File: lib/Dialect/SV/SVOps.cpp [hunk 2]
```diff
@@ -166,6 +166,33 @@ void IfDefOp::build(OpBuilder &odsBuilder, OperationState &result,
...
+  return llvm::all_of(block->without_terminator(),
+                      [](auto &op) { return false; });
```
Comment (lattner) [inline lib/Dialect/SV/SVOps.cpp:173]: For this, you can do something like:
`return block->empty() || block->front().hasTrait<OpTrait::IsTerminator>();`

File: lib/Dialect/SV/SVOps.cpp [hunk 3]
```diff
@@ -166,6 +166,33 @@ void IfDefOp::build(OpBuilder &odsBuilder, OperationState &result,
...
+void IfDefOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+                                          MLIRContext *context) {
+  // If both thenRegion and elseRegion are empty, erase op.
+  struct EraseEmptyOp final : public OpRewritePattern<IfDefOp> {
+    using OpRewritePattern::OpRewritePattern;
+    LogicalResult matchAndRewrite(IfDefOp op,
+                                  PatternRewriter &rewriter) const override {
+      if (!isEmptyBlockExceptForTerminator(op.getThenBlock()))
+        return failure();
+      else if (op.hasElse() &&
```
Comment (lattner) [inline lib/Dialect/SV/SVOps.cpp:185]: nit: no need for 'else' after return.

File: lib/Dialect/SV/SVOps.cpp [hunk 4]
```diff
@@ -236,7 +263,24 @@ void IfOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
...
     }
   };
+
+  struct EraseEmptyOp final : public OpRewritePattern<IfOp> {
+    using OpRewritePattern::OpRewritePattern;
+    LogicalResult matchAndRewrite(IfOp op,
+                                  PatternRewriter &rewriter) const override {
+      if (!isEmptyBlockExceptForTerminator(op.getThenBlock()))
+        return failure();
+      else if (op.hasElse() &&
```
Comment (lattner) [inline lib/Dialect/SV/SVOps.cpp:273]: same nit

File: lib/Dialect/SV/SVOps.cpp [hunk 5]
```diff
@@ -322,6 +366,23 @@ static void printEventList(OpAsmPrinter &p, AlwaysOp op, ArrayAttr portsAttr,
...
 }
 
+void AlwaysOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+                                           MLIRContext *context) {
+  // If the body is empty, erase op.
+  struct EraseEmptyOp final : public OpRewritePattern<AlwaysOp> {
+    using OpRewritePattern::OpRewritePattern;
+    LogicalResult matchAndRewrite(AlwaysOp op,
+                                  PatternRewriter &rewriter) const override {
+      if (!op.body().empty())
```
Comment (lattner) [inline lib/Dialect/SV/SVOps.cpp:376]: Should this use isEmptyBlockExceptForTerminator?  It has a terminator too IIRC.

File: lib/Dialect/SV/SVOps.cpp [hunk 6]
```diff
@@ -390,6 +451,22 @@ void AlwaysFFOp::build(OpBuilder &odsBuilder, OperationState &result,
...
+                                            MLIRContext *context) {
+  // If the body is empty, erase op.
+  struct EraseEmptyOp final : public OpRewritePattern<InitialOp> {
+    using OpRewritePattern::OpRewritePattern;
+    LogicalResult matchAndRewrite(InitialOp op,
+                                  PatternRewriter &rewriter) const override {
+      if (!op.body().empty())
+        return failure();
+
+      rewriter.eraseOp(op);
```
Comment (lattner) [inline lib/Dialect/SV/SVOps.cpp:464]: same

File: test/Dialect/SV/canonicalization.mlir [hunk 7]
```diff
@@ -43,5 +43,13 @@ func @if_dead_condition(%arg0: i1) {
...
   }
+
+  sv.if %arg0 {}
+  sv.if %arg0 {} else {}
+  sv.ifdef "SYNTHESIS" {}
+  sv.ifdef "SYNTHESIS" {} else {}
+  sv.always posedge %arg0 {}
+  sv.initial {}
+
   return
```
Comment (lattner) [inline test/Dialect/SV/canonicalization.mlir:54]: Test needs some check or CHECK-NOT: lines

**Review by lattner — APPROVED** [review]
Thanks for taking this on!

File: lib/Dialect/SV/SVOps.cpp (hunk 5 above)
Comment (uenoku) [reply lib/Dialect/SV/SVOps.cpp]: It looks `AlwaysOp` and `InitialOp` have been deleted without this PR so I have removed it

**Comment by lattner** [comment]
Looks great!

**Pull request merged**